core/src/proto/types.rs

```diff
@@ -1346,68 +1375,73 @@ impl Tx {
+                // signs over are present.
+                if hashes.iter().all(|x| {
                     signatures.targets.contains(x) || section.get_hash() == *x
-                }) {
-                    return Err(Error::InvalidSectionSignature(
-                        "missing target hash.".to_string(),
-                    ));
-                }
-
-                for target in &signatures.targets {
-                    if self.get_section(target).is_none() {
+                }) && signatures
+                    .targets
+                    .iter()
+                    .all(|x| self.get_section(x).is_some())
+                {
+                    if signatures.total_signatures() > max_signatures {
                         return Err(Error::InvalidSectionSignature(
-                            "Missing target section.".to_string(),
+                            "too many signatures.".to_string(),
                         ));
                     }
-                }
 
-                if signatures.total_signatures() > max_signatures {
-                    return Err(Error::InvalidSectionSignature(
-                        "too many signatures.".to_string(),
-                    ));
-                }
-
-                if signatures.total_signatures() < threshold {
-                    return Err(Error::InvalidSectionSignature(
-                        "too few signatures.".to_string(),
-                    ));
-                }
-
-                for signature_index in &signatures.signatures {
-                    let is_valid_signature = signature_index
-                        .verify(
+                    // Finally verify that the signature itself is valid
+                    let prev_verifieds = verified_pks.len();
+                    let amt_verifieds = signatures
+                        .verify_signature(
+                            &mut verified_pks,
                             &public_keys_index_map,
-                            &signatures.get_raw_hash(),
+                            signer,
                         )
-                        .is_ok();
-                    gas_meter
-                        .consume(VERIFY_TX_SIG_GAS_COST)
-                        .map_err(|_| Error::OutOfGas)?;
-                    if is_valid_signature {
-                        valid_signatures += 1;
+                        .map_err(|_| {
+                            Error::InvalidSectionSignature(
+                                "found invalid signature.".to_string(),
+                            )
+                        });
+                    // Compute the cost of the signature verifications
+                    if let Some(x) = gas_meter.as_mut() {
+                        let amt_verified = usize::from(amt_verifieds.is_err())
+                            + verified_pks.len()
+                            - prev_verifieds;
+                        x.consume(VERIFY_TX_SIG_GAS_COST * amt_verified as u64)
```

Comment:
I think that given this rework it would be slightly better to move the gas accounting part directly into the `verify_signature` method of `Signature`, to have a highest possible resolution. But we can do this later